### PR TITLE
Leaves - Raisah and Tiffany

### PR DESCRIPTION
**Pull request by TiffanyChio**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? | It seemed that the lessons following the introduction of the project were a big help in understanding the existing codebase. For example the lessons on abstract classes, template methods, etc. made it easier to understand the existing codebase. Maybe next time we want to read ahead on the lessons.  
What _inheritance_ relations exist between classes? | Passenger, Driver, and Trip are child classes of CsvRecord so they inherit methods and instance variables from CsvRecord. 
What _composition_ relations exist between classes? | TripDispatcher has many Trips, Drivers, and Passengers. Passenger objects have Trips. Drivers also have Trips. Trips have both Passenger and Drivers. 
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | One decision we made pertained to the driver fees associated with each trip. $1.65 is subtracted from the driver's total revenue per trip and we needed to decide what to do when trips cost less than $1.65. We could have deducted $1.65 per trip regardless of cost, deducted up to the total cost of the trip if the trip cost less than $1.65, or waived the fee. In the end we decided to waive the fee because it seemed the most humane and we are an ethical rideshare company. 
Give an example of a _template method_ that you implemented for this assignment | We didn't create a template method but there were existing template methods in CsvRecord. One template method that was used for the Trip, Passenger and Driver classes was "from_csv". "from_csv" had to be implemented differently for each class based on the parameters for instantiation.
Give an example of a _nominal test_ that you wrote for this assignment | We wrote a test for the Passenger "net_expenditures" method in which we passed in valid trip data for a Passenger object and tested if the trip costs were totaled through our "net_expenditures" method. 
Give an example of an _edge case test_ that you wrote for this assignment | We wrote a test to see if the Driver "total_revenue" method correctly calculates the driver's total revenue if trip cost is less than $1.65. As we described above, in this case, the driver should not be charged a fee since the cost of the trip is less than $1.65.
What is a concept that you gained more clarity on as you worked on this assignment | We gained more clarity on how abstract classes and template methods are used. And we also gained more clarity on the structure of test files. For example, we learned that if we instantiate an object that should throw an error but it is not encapsulated in the "expect" notation, then our test will not succeed. This is because Ruby will throw an error before we reach the "expect" block.

### CODE REVIEW

File: lib/passenger.rb [hunk 1]
```diff
@@ -3,21 +3,39 @@
...
     def add_trip(trip)
       @trips << trip
     end
-
+    
+    def net_expenditures
+      return 0 if @trips == []
+      
+      total = 0
+      @trips.each { |trip| total += trip.cost unless trip.end_time == nil }
```
Comment (beccaelenzil) [inline lib/passenger.rb:23]: You could consider using the sum enumerable (https://ruby-doc.org/core-2.4.0/Enumerable.html#method-i-sum).

File: test/trip_test.rb [hunk 2]
```diff
@@ -39,5 +46,63 @@
...
+          rating: 3,
+          driver_id: 123
+        )
+      }.must_raise ArgumentError
+    end
+  end
+  
+  describe "duration method" do
+    it "returns the correct duration for a trip in seconds" do
+      dur_test = RideShare::Trip.new(
```
Comment (beccaelenzil) [inline test/trip_test.rb:77]: Consider giving the variable `dur_test` and name like `trip_duration_test` to make it even more readable.

File: test/passenger_test.rb [hunk 3]
```diff
@@ -1,74 +1,174 @@
...
+    
     it "all Trips must have the same passenger's passenger id" do
       @passenger.trips.each do |trip|
         expect(trip.passenger.id).must_equal 9
       end
     end
   end
-
+  
   describe "net_expenditures" do
```
Comment (beccaelenzil) [inline test/passenger_test.rb:71]: I think you're missing an edge case for both net_expenditures and total_time_spent. What happens if the user has no trips?

File: lib/driver.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    
+    def change_status_to_unavailable
+      @status = :UNAVAILABLE
+    end
+    
+    def average_rating
+      return 0 if @trips == []
+      
+      # Excludes trips with end time nil
+      completed_trips = @trips.select {|trip| trip.end_time != nil}
```
Comment (beccaelenzil) [inline lib/driver.rb:30]: Great use of enumerables.

File: test/driver_test.rb [hunk 5]
```diff
@@ -114,23 +114,135 @@
...
+      driver = RideShare::Driver.new(
+        id: 54,
+        name: "Rogers Bartell IV",
+        vin: "1C9EVBRM0YBC564DZ"
+      )
+      expect(driver.total_revenue).must_equal 0
+    end
+    
+    it "correctly calculates the total revenue" do
+      trip2 = RideShare::Trip.new(
```
Comment (beccaelenzil) [inline test/driver_test.rb:180]: Instead of `trip2`, consider naming this variable `trip` in each test for consistency.

**Comment by beccaelenzil** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | check
Answer comprehension questions | check
**Wave 1** |
Appropriate use of Ruby's `Time` | check
`Trip` has a helper method to calculate duration | check
`Passenger` methods to calculate total cost and total time | check
Complete tests for new behavior | check -- see comment for testing edge case
**Wave 2** |
`Driver` inherits from `CsvRecord`, and implements `from_csv` | check
`Driver`s and `Trip`s are connected by `TripDispatcher` | check
`Driver` has method to calculate average rating | check
`Driver` has a method to calculate total revenue | check
**Wave 3** |
`TripDispatcher` has a new method to create trips | check
Trip creation logic is correctly implemented | check
New `Trip` is connected to the appropriate `Driver` and `Passenger` | check
Tests for `request_trip` | check
Wave 1 and 2 methods ignore in-progress trips | check
Tests for Wave 1 and 2 methods w/ in-progress trips | check
**Wave 4 (Optional)** |
`TripDispatcher` assigns trips to the correct driver | check
Complete tests | check
Appropriate helper methods were made to help with complex logic | checl
**Overall** | Good job overall! Your code is well-organized and easy to read, and does a good job of solving the problem at hand. I've left a few inline comments for you to review, but it's clear that the learning goals for this assignment were met. You make a good use of comments to increase readability where appropriate. Great work implementing the additional wave. Keep up the hard work!
